Feature/controller manager

Comerm28 opened the pull request:
wasnt sure if i should have included my cmake chanegs to build my tests, I did but commented them out, can remove if needed

RCMast3r committed:
added counter to enum

RCMast3r committed:
added config handling for state estimator

RCMast3r committed:
added comments

RCMast3r committed:
adding more doxygen comments and comments in general

Comerm28 committed:
temp pausing, swapping controllers, controllermanagerstate struct

Comerm28 committed:
start of documentation, switch controller logic implemented -> need to test

Comerm28 committed:
removed state estimator variable and added more comments

Comerm28 committed:
for rebase to get controller code

Comerm28 committed:
passes tests

Comerm28 committed:
state estimator shouldnt show in pr

Comerm28 committed:
fixed commit problems

Comerm28 committed:
resetting more unintended rebase changes

RCMast3r commented on test/test_controller_manager.cpp:
needs more tests between real controllers outputs. also check to see what happens when we swap between a "speed" controller with speed setpoints being wither 0 or max rpm and varying the torque to a regular torque controller.

Comerm28 committed:
trying to merge

Comerm28 committed:
Merge branch 'master' into feature/controller-manager

KrishKittur commented on drivebrain_core_base/include/public/ControllerManager.tpp:
@RCMast3r Is there a reason that the controller manager is implemented with templates? I think it could make more sense architecturally (and also make the code more maintainable) if methods like `_can_switch_controller` were virtual functions within the `Controller` class with the necessary inputs. And, the `ControllerManager` would just take an array of these controllers, and a state machine (sorta like what is there right now) could invoke the necessary methods of the Controllers via polymorphism to handle state transitions.

RCMast3r replied:
it is implemented with templates due to the size of the array within the controller manager being determined by passed-through template arguments so we can compose the controller manager more easily externally / on construction. due to separation of concerns, the controllers themselves should be focused on control of the car whereas the controller manager is dedicated to just switching controllers. the controllers themselves shouldnt care about whether or not they themselves can be switched between as to them they are just creating control commands / dedicated to calculating control commands. especially when it comes to independent testing as each controller could have its own implementation of the controller muxing logic.

KrishKittur replied:
Ah makes sense thanks.

Comerm28 committed:
one test

Comerm28 committed:
test
# Please enter the commit message for your changes. Lines starting
# with '#' will be ignored, and an empty message aborts the commit.
#
# On branch feature/controller-manager
# Your branch is up to date with 'origin/feature/controller-manager'.
#
# Changes to be committed:
#	modified:   ../drivebrain_app/main.cpp
#	modified:   ../drivebrain_core_base/include/public/ControllerManager.tpp
#	modified:   ../drivebrain_core_impl/drivebrain_control/src/SimpleController.cpp
#	modified:   ../test/test_controller_manager.cpp
#

Comerm28 committed:
Merge branch 'master' into feature/controller-manager

Comerm28 committed:
build tyoe hit

Comerm28 committed:
fixed config issues and added logger

Comerm28 committed:
state_estimator ptr

Comerm28 committed:
one more test case

Comerm28 committed:
no switch on same ctr

Comerm28 committed:
Merge branch 'master' into feature/controller-manager

Comerm28 committed:
merge plus test = ctr manager

Comerm28 committed:
stuff and things

Comerm28 committed:
built locally

Comerm28 committed:
config

Comerm28 committed:
minor test errors

Comerm28 committed:
forgot to set brake to 0

RCMast3r commented on drivebrain_core_impl/drivebrain_comms/include/DBServiceImpl.hpp:
```diff
@@ -15,14 +18,17 @@ class DBInterfaceImpl final : public db_service::v1::service::DBInterface::Servi
         void stop_server();
     private:
         std::shared_ptr<core::MsgLogger<std::shared_ptr<google::protobuf::Message>>> _logger_inst;
+        std::shared_ptr<control::ControllerManager<control::Controller<core::ControllerOutput, core::VehicleState>, 2>> _ctr_manager_inst;
```
https://github.com/hytech-racing/drivebrain_software/blob/master/drivebrain_app/main.cpp#L114 this is how to register the member functions

implementation of msglogger:

https://github.com/hytech-racing/drivebrain_software/blob/master/drivebrain_core_base/include/public/MsgLogger.hpp#L30

Comerm28 replied:
switching to a vector makes sense to not hard code controller number, but if we pass in funcs instead of the controllers, would there not need to be multiple vectors for each individual function that the controllers need to be called on, right now I would need one vector for the get dt secs and one for the stepping(meaning if more functions are added to the controller then I would need more and more vectors), does it not make more sense to have a vector of the controllers themselves? I could also be missing something about what u mean by the function, but the msglogger seems to be using much more singular, standalone methods

RCMast3r replied:
> switching to a vector makes sense to not hard code controller number, but if we pass in funcs instead of the controllers, would there not need to be multiple vectors for each individual function that the controllers need to be called on, right now I would need one vector for the get dt secs and one for the stepping(meaning if more functions are added to the controller then I would need more and more vectors), does it not make more sense to have a vector of the controllers themselves? I could also be missing something about what u mean by the function, but the msglogger seems to be using much more singular, standalone methods

I would prefer that we did not use a vector of controller numbers. we do not change the number of controllers at runtime which would be a reason to use a vector.

A template argument is not a "hard coded" value when it comes to this.

also when it comes to the passing of funcs i am talking about using them to pass in the `swap_active_controller` member function from the controller manager. im not quite sure what you are referring to with the multiple vectors for each individual function.

Comerm28 replied:
misunderstood this to mean remove the template and pass in funcs to the ctr manager, but alr did that, just got it fully unpointered

RCMast3r commented on drivebrain_core_impl/drivebrain_comms/include/DBServiceImpl.hpp:
```diff
@@ -15,14 +18,17 @@ class DBInterfaceImpl final : public db_service::v1::service::DBInterface::Servi
     private:
         std::shared_ptr<core::MsgLogger<std::shared_ptr<google::protobuf::Message>>> _logger_inst;
+        std::shared_ptr<control::ControllerManager<control::Controller<core::ControllerOutput, core::VehicleState>, 2>> _ctr_manager_inst;
+        core::StateEstimator* _state_estimator;
```
I see that you are using the state estimator here:

https://github.com/hytech-racing/drivebrain_software/blob/51807048926687502ebbf4677b71d909aa57ad50/drivebrain_core_impl/drivebrain_comms/src/DBServiceImpl.cpp#L39

Comerm28 committed:
commit for nix

Comerm28 committed:
include issue

Comerm28 committed:
Merge branch 'master' into feature/controller-manager

Comerm28 committed:
perchance

Comerm28 committed:
used wrong ctr manager in some tests

Comerm28 committed:
i misunderstood but it should be good now assuming tests still pass

Comerm28 committed:
Merge branch 'master' into feature/controller-manager

RCMast3r commented:
is now a part of `dev/v1.1.0`
